tsl/src/fdw/data_node_scan_plan.c

```diff
@@ -741,6 +755,691 @@ push_down_group_bys(PlannerInfo *root, RelOptInfo *hyper_rel, Hyperspace *hs,
+			 * the fpinfo of this relation so that the deparser can take
+			 * appropriate action.  Also, save the relids of base relations
+			 * covered by that relation for later use by the deparser.
+			 */
+			if (fpinfo_o->remote_conds)
+			{
+				fpinfo->make_outerrel_subquery = true;
+				fpinfo->lower_subquery_rels =
+					bms_add_members(fpinfo->lower_subquery_rels, outerrel->relids);
+			}
+			if (fpinfo_i->remote_conds)
+			{
+				fpinfo->make_innerrel_subquery = true;
+				fpinfo->lower_subquery_rels =
+					bms_add_members(fpinfo->lower_subquery_rels, innerrel->relids);
+			}
+			break;
+
+		default:
+			/* Should not happen, we have just checked this above */
+			elog(ERROR, "unsupported join type %d", jointype);
+	}
+
+	/*
+	 * For an inner join, all restrictions can be treated alike. Treating the
+	 * pushed down conditions as join conditions allows a top level full outer
+	 * join to be deparsed without requiring subqueries.
+	 */
+	if (jointype == JOIN_INNER)
+	{
+		Assert(!fpinfo->joinclauses);
+		fpinfo->joinclauses = fpinfo->remote_conds;
+		fpinfo->remote_conds = NIL;
+	}
+	/* Mark that this join can be pushed down safely */
+	fpinfo->pushdown_safe = true;
+
+	/*
+	 * Set the string describing this join relation to be used in EXPLAIN
+	 * output of corresponding ForeignScan.  Note that the decoration we add
+	 * to the base relation names mustn't include any digits, or it'll confuse
+	 * postgresExplainForeignScan.
+	 */
+	fpinfo->relation_name = makeStringInfo();
+	appendStringInfo(fpinfo->relation_name,
+					 "(%s) %s JOIN (%s)",
+					 fpinfo_o->relation_name->data,
+					 get_jointype_name(fpinfo->jointype),
+					 fpinfo_i->relation_name->data);
+
+	/*
+	 * Set the relation index.  This is defined as the position of this
+	 * joinrel in the join_rel_list list plus the length of the rtable list.
+	 * Note that since this joinrel is at the end of the join_rel_list list
+	 * when we are called, we can get the position by list_length.
+	 */
+	fpinfo->relation_index = list_length(root->parse->rtable) + list_length(root->join_rel_list);
+
+	return true;
+}
+
+#ifdef USE_ASSERT_CHECKING
+/*
+ * Check if the given hypertable is a distributed hypertable.
+ */
+static bool
+is_distributed_hypertable(Oid hypertable_id)
+{
+	Cache *hcache;
+
+	Hypertable *ht =
+		ts_hypertable_cache_get_cache_and_entry(hypertable_id, CACHE_FLAG_MISSING_OK, &hcache);
+
+	/* perform check before cache is released */
+	bool ht_is_distributed = (ht != NULL && hypertable_is_distributed(ht));
+	ts_cache_release(hcache);
+
+	return ht_is_distributed;
+}
+#endif
+
+/*
+ * Create a new join partition RelOptInfo data structure for a partition. The data
+ * structure is based on the parameter joinrel. The paramter is taken as template
+ * and adjusted for the partition provided by the parameter hyper_table_rel.
+ */
+static RelOptInfo *
+create_join_partition(PlannerInfo *root, RelOptInfo *innerrel, RelOptInfo *joinrel,
```

Comment:
`create_join_partition_for_data_node`? Or `create_data_node_partition_for_join` :thinking:  Just to make it sound less generic, we have other kinds of partition-wise joins.

Reply:
`create_data_node_partition_for_joinrel` is even more specific.

Reply:
I only just realized that this is exactly the same kind of partitioning we have for data node scans :) Somehow the word "partition" confused me into thinking about generic partition-wise joins, that would be chunk-wise in our terms. Maybe `data_node_joinrel` would be even better...

Reply:
The function has been renamed to `create_data_node_joinrel`.